src/sunmatrix/sparse/sunmatrix_sparse.c

```diff
@@ -656,30 +661,44 @@ int SUNMatScaleAddI_Sparse(realtype c, SUNMatrix A)
     /* iterate through columns (rows) backwards */
     for (j=N-1; j>=0; j--) {
 
-      /* clear out temporary arrays for this column (row) */
-      for (i=0; i<M; i++) {
-        w[i] = 0;
-        x[i] = RCONST(0.0);
-      }
+      /* clear out diagonal element for this column (row) */
+      x[j] = RCONST(0.0);
 
       /* iterate down column (row) of A, collecting nonzeros */
+      i = 0;
       for (p=Ap[j]; p<cend; p++) {
-        w[Ai[p]] += 1;         /* indicate that row (column) is filled */
+        w[i++] = Ai[p];        /* collect row (column) index */
         x[Ai[p]] = c*Ax[p];    /* collect/scale value */
       }
 
+      /* NNZ in this column (row) */
+      nw = cend - Ap[j];
+
       /* add identity to this column (row) */
       if (j < M) {
-        w[j] += 1;     /* indicate that row (column) is filled */
         x[j] += ONE;   /* update value */
       }
 
       /* fill entries of A with this column's (row's) data */
-      for (i=M-1; i>=0; i--) {
-        if ( w[i] > 0 ) {
-          Ai[--nz] = i;
-          Ax[nz] = x[i];
-        }
+      for (i=nw-1; i>=0 && w[i]>j; i--) {
+        Ai[--nz] = w[i];
+        Ax[nz] = x[w[i]];
+      }
+      if (w[i] != j) {
+        Ai[--nz] = j;
+        Ax[nz] = x[j];
+      }
+      for (; i>=0; i--) {
+        Ai[--nz] = w[i];
+        Ax[nz] = x[w[i]];
+      }
+
+      /* clear out temporary values for this column (row) */
+      for (i=0; i<nw; i++) {
+        x[w[i]] = RCONST(0.0);
+      }
+      if (j < M) {
+        x[j] = RCONST(0.0);
       }
```

Comment:
Is this block even necessary?  It looks like the above code only ever uses values from `x` that it has filled in, so pre-existing values are just ignored.

Reply:
That's right